src/ledger/LedgerStateSnapshot.h

```diff
@@ -0,0 +1,179 @@
+    virtual LedgerHeaderWrapper getHeader() = 0;
+    virtual LedgerEntryWrapper getSourceAccount(AccountID const& account) = 0;
+    virtual LedgerEntryWrapper
+    getSourceAccount(LedgerHeaderWrapper const& header,
+                     TransactionFrame const& tx) = 0;
+    virtual LedgerEntryWrapper
+    getSourceAccount(LedgerHeaderWrapper const& header,
+                     TransactionFrame const& tx,
+                     AccountID const& AccountID) = 0;
+    virtual LedgerEntryWrapper load(LedgerKey const& key) = 0;
+#ifdef BUILD_TESTS
+    virtual LedgerEntryWrapper loadToModify(LedgerKey const& key) = 0;
+#endif
+};
+
+// A concrete implementation of read-only SQL snapshot wrapper
+class LedgerTxnReadOnly : public AbstractLedgerStateSnapshot
+{
+    // Callers are expected to manage `AbstractLedgerTxn` themselves.
+    // LedgerTxnReadOnly guarantees that LedgerTxn, LedgerTxnHeader and
+    // LedgerTxnEntry referenced all remain valid.
+    AbstractLedgerTxn& mLedgerTxn;
+
+  public:
+    LedgerTxnReadOnly(AbstractLedgerTxn& ltx);
+    ~LedgerTxnReadOnly() override;
+    LedgerHeaderWrapper getLedgerHeader() override;
+    LedgerHeaderWrapper getHeader() override;
+    LedgerEntryWrapper getSourceAccount(AccountID const& account) override;
+    LedgerEntryWrapper getSourceAccount(LedgerHeaderWrapper const& header,
+                                        TransactionFrame const& tx) override;
+    LedgerEntryWrapper getSourceAccount(LedgerHeaderWrapper const& header,
+                                        TransactionFrame const& tx,
+                                        AccountID const& AccountID) override;
+    LedgerEntryWrapper load(LedgerKey const& key) override;
+#ifdef BUILD_TESTS
+    LedgerEntryWrapper loadToModify(LedgerKey const& key) override;
+#endif
+    AbstractLedgerTxn&
+    getLedgerTxn() const
+    {
+        return mLedgerTxn;
+    }
+};
+
+// A concrete implementation of read-only BucketList snapshot wrapper
+class BucketSnapshotState : public AbstractLedgerStateSnapshot
+{
+    BucketManager& mBucketManager;
+    std::shared_ptr<SearchableBucketListSnapshot> mSnapshot;
+
+  public:
+    BucketSnapshotState(BucketManager& bm);
+    ~BucketSnapshotState() override;
+
+    LedgerHeaderWrapper getLedgerHeader() override;
+    LedgerHeaderWrapper getHeader() override;
+    LedgerEntryWrapper getSourceAccount(AccountID const& account) override;
+    LedgerEntryWrapper getSourceAccount(LedgerHeaderWrapper const& header,
+                                        TransactionFrame const& tx) override;
+    LedgerEntryWrapper getSourceAccount(LedgerHeaderWrapper const& header,
+                                        TransactionFrame const& tx,
+                                        AccountID const& AccountID) override;
+    LedgerEntryWrapper load(LedgerKey const& key) override;
+#ifdef BUILD_TESTS
+    LedgerEntryWrapper
+    loadToModify(LedgerKey const& key) override
```

Comment:
This should never be called for a `BucketList` returned shared_ptr. I would either throw, or since we only use it in loadgen, remove it to just the `LedgerTxnReadOnly` and static_cast when necessary,